### ISSUE TITLE
Getting Prediction Values Change for individual instances

### ISSUE THREAD

**Issue by iteimouri** [catboost/catboost#1538]
I have trained a model for a binary classification. I am interested in finding the Prediction Value Change for individual instances. I know that using 

```
model.get_feature_importance(Pool(X_test, label=y_test.,cat_features=categorical_features_indices),
type='PredictionValuesChange')
```
Gives me the values over the whole test set. I am trying to do this for individual instance such as: 

```
model.get_feature_importance(Pool(X_test.iloc[0:1], label=y_test.iloc[0:1],cat_features=categorical_features_indices), 
                                                                     type='PredictionValuesChange')
```
Despite the fact that `Pool()` returns as object, it seems get_feature_importance() method is not happy with individual rows. This is somehow would be similar to `shap.force_plot()` method where one can see the SHAP values on each instances. I wonder if making `PredictionValuesChange` makes sense for individual cases at all?

**Comment by ljluestc** [comment]
```

import numpy as np
import pandas as pd
from catboost import CatBoostClassifier, Pool
from sklearn.datasets import make_classification
from sklearn.model_selection import train_test_split
import shap

# Generate sample data
X, y = make_classification(n_samples=1000, n_features=10, n_classes=2, random_state=42)
X = pd.DataFrame(X, columns=[f'feature_{i}' for i in range(X.shape[1])])
categorical_features_indices = []  # Empty list since our example has no categorical features

# Split data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Train CatBoost model
model = CatBoostClassifier(
    iterations=100,
    learning_rate=0.1,
    depth=6,
    random_seed=42,
    verbose=0
)
model.fit(X_train, y_train, cat_features=categorical_features_indices)

# Approach 1: Using CatBoost's built-in SHAP values for individual instances
def get_individual_shap_values(model, X_instance):
    """Get SHAP values for a single instance using CatBoost's built-in method"""
    pool_instance = Pool(X_instance, cat_features=categorical_features_indices)
    shap_values = model.get_feature_importance(
        data=pool_instance,
        type='ShapValues',  # Note: different from PredictionValuesChange
        prettified=False
    )
    # shap_values shape: (n_samples, n_features + 1), last column is base value
    return shap_values[0, :-1], shap_values[0, -1]  # Return feature contributions and base value

# Approach 2: Custom approximation of PredictionValuesChange for single instance
def get_individual_prediction_change(model, X_instance, X_reference, n_permutations=10):
    """Approximate feature importance for a single instance by comparing to reference set"""
    pool_instance = Pool(X_instance, cat_features=categorical_features_indices)
    base_pred = model.predict_proba(pool_instance)[0]
    
    feature_changes = {}
    n_features = X_instance.shape[1]
    
    for feature_idx in range(n_features):
        total_change = 0
        feature_name = X_instance.columns[feature_idx]
        
        for _ in range(n_permutations):
            # Create modified instance by replacing feature value with reference set value
            X_modified = X_instance.copy()
            ref_value = X_reference.iloc[np.random.randint(0, len(X_reference))][feature_name]
            X_modified.iloc[0, feature_idx] = ref_value
            
            pool_modified = Pool(X_modified, cat_features=categorical_features_indices)
            modified_pred = model.predict_proba(pool_modified)[0]
            
            # Calculate change in prediction probability (using absolute difference)
            change = np.abs(base_pred - modified_pred).sum()
            total_change += change
        
        feature_changes[feature_name] = total_change / n_permutations
    
    return feature_changes

# Example usage
print("=== Testing with single instance ===")
instance_idx = 0
X_instance = X_test.iloc[instance_idx:instance_idx+1]
true_label = y_test[instance_idx]

# Get prediction
pool_instance = Pool(X_instance, cat_features=categorical_features_indices)
prediction = model.predict(pool_instance)[0]
print(f"Instance {instance_idx} - True label: {true_label}, Predicted: {prediction}")

# Approach 1: SHAP values
shap_contributions, base_value = get_individual_shap_values(model, X_instance)
print("\nSHAP Values (recommended):")
for feature, value in zip(X_instance.columns, shap_contributions):
    print(f"{feature}: {value:.4f}")
print(f"Base value: {base_value:.4f}")

# Approach 2: Custom Prediction Change
pred_changes = get_individual_prediction_change(model, X_instance, X_test)
print("\nCustom Prediction Value Changes (approximation):")
for feature, change in pred_changes.items():
    print(f"{feature}: {change:.4f}")

# Visualize SHAP force plot (using shap library)
explainer = shap.TreeExplainer(model)
shap_values = explainer.shap_values(X_test)
print("\nGenerating SHAP force plot...")
shap.force_plot(
    explainer.expected_value[1],
    shap_values[1][instance_idx],
    X_test.iloc[instance_idx],
    matplotlib=True,
    show=True
)

# Compare with dataset-wide PredictionValuesChange
print("\nDataset-wide PredictionValuesChange (for reference):")
test_pool = Pool(X_test, y_test, cat_features=categorical_features_indices)
feature_importance = model.get_feature_importance(
    data=test_pool,
    type='PredictionValuesChange',
    prettified=True
)
print(feature_importance)
​​​​​​​​​​​​​​​​​​​​​​​​​​​​​​​​​​​​​​​​​​​​​​​​​​
```
